parser/src/ast.rs

```diff
@@ -0,0 +1,101 @@
+use crate::lexer::Token;
+use std::fmt;
+
+/// [Identifier] is used to represent variable names.
+#[derive(Debug, Eq, PartialEq)]
+pub struct Identifier {
+    pub name: String,
+}
+
+impl fmt::Display for Identifier {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "{}", &self.name)
+    }
+}
+
+/// [Source] is the root node of the AST representing the AIR constraints file.
+#[derive(Debug, PartialEq)]
+pub struct Source(pub Vec<SourceSection>);
+
+/// Source is divided into SourceSections.
+/// There are 4 types of Source Sections:
+/// - AirDef: Name of the air constraints module.
+/// - TraceCols: Trace Columns representing columns of the execution trace.
+/// - BoundaryConstraints: Boundary Constraints to be enforced on the boundaries of columns defined
+///   in the TraceCols section. Currently there are two types of boundaries, First and Last
+///   representing the first and last rows of the column.
+/// - TransitionConstraints: Transition Constraints to be enforced on the trace columns defined
+///   in the TraceCols section.
+#[derive(Debug, PartialEq)]
+pub enum SourceSection {
+    AirDef(AirDef),
+    TraceCols(TraceCols),
+    BoundaryConstraints(BoundaryConstraints),
+    TransitionConstraints(TransitionConstraints),
+}
+
+/// Name of the air constraints module.
+#[derive(Debug, Eq, PartialEq)]
+pub struct AirDef {
+    pub name: Identifier,
+}
+
+/// [TraceCols] contains the trace columns representing columns of the execution trace.
+/// TraceCols is made up of TraceColsGrp which represents groups of columns of a specific type.
+#[derive(Debug, PartialEq)]
+pub struct TraceCols {
+    pub cols: Vec<TraceColsGrp>,
+}
+
+/// [TraceColsGrp] represents a group of columns of a specfic type.
+/// There are two types of groups:
+/// - MainTraceCols
+/// - AuxiliaryTraceCols
+#[derive(Debug, PartialEq)]
+pub struct TraceColsGrp {
+    pub cols_grp_type: TraceColsGrpType,
+    pub cols: Vec<Expr>,
```

Comment:
Does this imply that `cols` could contain any expressions? If so, should we limit it as there is only very restricted set of options which could be used here? Specifically, for this PR we could restrict this only to a list of identifiers. For example:
```
trace_columns:
    main: [a, b, c, d]
```
In the future PRs, we could add more complex syntax. For example:
```
trace_columns:
    main: [a, b[2], c[3], d, ..5]
```
Or something like that - but we should probably discuss first what syntax we want to allow.

Reply:
Makes sense. Since this has been changed to what you suggested above now, main and auxiliary columns can only contain identifiers for now.